Update file targeting for the second run of inter-file diff scan

ihji opened the pull request:
Scan the same set of files for both the first and second runs of the differential scan to remove duplicated findings.

Closes SAF-971

test plan: https://github.com/semgrep/semgrep-proprietary/pull/1407

IagoAbal commented on cli/src/semgrep/scan_report.py:
```diff
@@ -116,7 +116,7 @@ def _print_scan_plan_header(
         )
         diff_file_count = len(target_mode_config.get_diff_targets())
-        summary_line = f"Pro Differential Scanning {diff_file_count}/{unit_str(total_file_count, 'file')}"
+        summary_line = f"Pro Differential Scanning {unit_str(diff_file_count, 'file')}"
```
Why changing the message format here?

ihji replied:
I think showing only the number of changed files is less confusing, so I simplified it.

IagoAbal replied:
Perhaps it should be "Inter-file Differential ..." instead of "Pro Differential ..." ?

ihji replied:
done.

IagoAbal commented on src/reporting/Core_json_output.ml:
```diff
@@ -595,7 +595,7 @@ let core_output_of_matches_and_errors (res : Core_result.t) : OutJ.core_output =
          * but should we compute skipped and scanned here instead?
```
What about this `TODO` note?

ihji replied:
I'm not totally sure what it means. There are two paths of output due to `osemgrep` and `pysemgrep`. This path is only related to `pysemgrep` output. I assume that we need to update here too. `osemgrep` output already has the `scanned` data.

IagoAbal replied:
We need to at least update that TODO note or something, according to our best knowledge.

ihji replied:
done.

IagoAbal commented on cli/src/semgrep/run_scan.py:
```diff
@@ -633,13 +633,14 @@ def run_scan(
                     baseline_target_strings = target_strings
                     baseline_target_mode_config = target_mode_config
                     if target_mode_config.is_pro_diff_scan:
+                        scanned = [
```
Can we just pass the `scanned` list to the baseline scan? There may be files in there that do not exist in the baseline version. What happens if a file gets renamed, should we scan the same file under the old name?

ihji replied:
Added a renamed set of files to the target list. Deleted files are checked and removed by `t.exists()` below. A test for removing duplicated findings in renamed files has also been added to the Pro engine PR.

IagoAbal commented on cli/src/semgrep/run_scan.py:
```diff
@@ -635,13 +635,15 @@ def run_scan(
                     baseline_target_strings = target_strings
```
We should have a changelog.

ihji replied:
done.

IagoAbal commented on cli/src/semgrep/run_scan.py:
```diff
@@ -635,13 +635,15 @@ def run_scan(
                     baseline_target_strings = target_strings
                     baseline_target_mode_config = target_mode_config
```
We need the Pro PR to not be a draft one and be all green (except for the "check-submodule" workflow).

ihji replied:
done.

ihji committed:
Update file targeting for the second run of inter-file diff scan

Scan the same set of files for both the first and second runs of the
differential scan to remove duplicated findings.

ihji commented:
Windows build failure seems irrelevant.

IagoAbal commented on changelog.d/saf-897.fixed:
```diff
@@ -0,0 +1,3 @@
+Improved removal of pre-existing findings in inter-file diff scan: We've
```
I'd describe what was happening, "when renaming or adding new files, inter-file diff scans could report pre-existing findings". More than improving it, we've fixed the problem IMO.

ihji replied:
done.

IagoAbal commented on cli/src/semgrep/run_scan.py:
```diff
@@ -635,13 +635,15 @@ def run_scan(
                     baseline_target_strings = target_strings
                     baseline_target_mode_config = target_mode_config
                     if target_mode_config.is_pro_diff_scan:
+                        scanned = [
```
It would be nice having a comment here explaining what we do and why. We have to derive the set of files to scan the baseline from the previous scan because otherwise we may analyze fewer files, etc.

ihji replied:
done.

IagoAbal commented:
@ihji please wait for @emjin to take a second look, I'm not super familiar with this code and I'd appreciate that Emma reviews it too.

emjin commented on cli/src/semgrep/run_scan.py:
```diff
@@ -635,13 +635,15 @@ def run_scan(
                     baseline_target_strings = target_strings
                     baseline_target_mode_config = target_mode_config
                     if target_mode_config.is_pro_diff_scan:
```
Minor note that we should change references like these to "interfile_diff_scan"

ihji replied:
added a todo note.

emjin reviewed (approved):
Seems reasonable to me but I haven't actually interacted with this part of the code much

ihji committed:
update
